### ISSUE TITLE
AWS: Node becomes NodeNotReady without logged reason

### ISSUE THREAD

**Issue by willmore** [kubernetes/kubernetes#41916]
BUG REPORT

**Kubernetes version**:
`1.5.2`

**Environment**:
- **Cloud provider**: AWS
- **OS**: Container Linux by CoreOS 1235.9.0 (Ladybug)
- **Kernel**: 4.7.3-coreos-r2
- **Install tools**: kube-aws
- **Other notes**: Running HA masters - clustered etcd; API servers behind ELB.

**What happened**:
All nodes dropped themselves from the cluster, nearly simultaneously.
I.e., within the span of seconds, all nodes reported themselves to master with status `NodeNotReady`.
This happened in two independent clusters, installed in two different regions. 
There is no indication from any log (controller, kubelet, etc.) why the kubelet transitioned to `NodeNotReady`.

For each node, a entry similar to the following appeared in log of `kube-controller-manager`:
```
I0217 14:19:11.432691 1 event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"ip-XX-XX-XX-XX.ec2.internal", UID:"XXX", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node ip-XX-XX-XX-XX.ec2.internal status is now: NodeNotReady
```

I have encountered similar incidents in the past, in which all nodes of a AWS cluster would mark themselves not ready due to outages of an EC2 API (similar to [this](https://github.com/kubernetes/kubernetes/issues/41904)) - no indication of that here.

**What you expected to happen**:
I expected minimally, a clear indication as to _why_ a node transitions to `NodeNotReady`, both in the controller and kubelet logs.

**How to reproduce it**:
Cannot say, other than running a cluster with similar configuration on AWS.

**Comment by justinsb** [comment]
Given the timing, I'm suspicious that this is the same as #41904.  But we would expect to see the error in the kubelet logs, and that only affected one of my nodes.  I'd also wonder about a time change, but it's february so I don't think that's the case.

There's nothing in `kubectl describe nodes` either?

**Comment by willmore** [comment]
This event occurred in both our regions (us-east-1, eu-west-1) on 17. Feb (not same date as https://github.com/kubernetes/kubernetes/issues/41904).

describe nodes gives as example:
```
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  OutOfDisk 		False 	Wed, 22 Feb 2017 20:21:00 +0000 	Fri, 17 Feb 2017 14:34:18 +0000 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Wed, 22 Feb 2017 20:21:00 +0000 	Fri, 10 Feb 2017 12:07:13 +0000 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  DiskPressure 		False 	Wed, 22 Feb 2017 20:21:00 +0000 	Fri, 10 Feb 2017 12:07:13 +0000 	KubeletHasNoDiskPressure 	kubelet has no disk pressure
  Ready 		True 	Wed, 22 Feb 2017 20:21:00 +0000 	Fri, 17 Feb 2017 14:34:18 +0000 	KubeletReady 			kubelet is posting ready status
```

The entries with 17. Feb timestamp are when the nodes came back online (from off back to on was ~15 min.).

**Comment by willmore** [comment]
I will preemptively comment on the `LastTransitionTime` of `OutOfDisk`, whose timestamp matches the outage. Our disk monitoring shows that the disks across the cluster were in no way at risk of filling at the time.

**Comment by justinsb** [comment]
That is interesting though - it does look like OutOfDisk was the thing that signalled.  We also had recently reported https://github.com/kubernetes/kubernetes/issues/41850 cc @colemickens 

I can't imagine a scenario where disks being full would be correlated.  Not can I imagine a situation where it could be incorrectly reported in a correlated way.

I suspect this is the same root cause as #41850, and a different root cause from #41904

**Comment by willmore** [comment]
Interesting. Please let me know if there is anything we can scour logs for or monitor.

**Comment by deiwin** [comment]
I work with @willmore, and I've been going through the logs. The only interesting find from the kubelet logs is the fact that there's usually a high frequency (roughly 25 per minute for one node) of the following messages in the logs:

```
operation_executor.go:917] MountVolume.SetUp succeeded for volume "kubernetes.io/secret/[...]
```

But these messages stopped suddenly in every affected node ~30s before the Controller-Manager reported `NodeNotReady` for these nodes.

These messages return around the time when Node changes back to `Ready`. Initially with `no route to host` errors, but successfully shortly after. Sample of the inital errors after recovery:

```
secret.go:197] Couldn't get secret [...]
nestedpendingoperations.go:262] Operation for "\"kubernetes.io/secret/[...]\" (\"[...]\")" failed. No retries permitted until 2017-02-17 14:34:14.015072729 +0000 UTC (durationBeforeRetry 500ms). Error: MountVolume.SetUp failed for volume "kubernetes.io/secret/[...]\" (spec.Name: "[...]") pod "[...]" (UID: "[...]") with: Get https://kube-prod-us.service/api/v1/namespaces/kube-system/secrets/[...]: read tcp XX.XX.XX.XX:55060->XX.XX.XX.XX:443: read: no route to host
```

**Comment by jswoods** [comment]
I am also seeing this issue and I'm running 1.5.5. The LastTransitionTime for KubeletHasSufficientDisk aligns with the time the nodes re-joined the cluster, but the nodes did not have low disk space.

There is nothing notable in the kubelet with regard to the NodeNotReady state. But as with @deiwin, I do see a decrease in the `MountVolume.SetUp succeeded for volume` messages. I also saw this in the kubelet logs around the time the nodes went into a Ready state:

```
Couldn't get secret <secret>
```
and

```
kubelet_pods.go:610] Unable to retrieve pull secret <namespace>/registrypullsecret for <namespace/<pod> due to Get https://kubernetes/api/v1/namespaces/<namespace>/secrets/regis
trypullsecret: read tcp <ip>:38600-><ip>:443: read: connection reset by peer.  The image pull may not succeed.
```

Also perhaps somewhat obvious, this popped up in the controller logs:
```
event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"<node>", UID:"cd146421-13d7-11e7-b5f5-060a60429fa6", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node <node> status is now: NodeNotReady
event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"<node>", UID:"c913fe34-13d7-11e7-b5f5-060a60429fa6", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node <node> status is now: NodeNotReady
event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"<node>", UID:"d0027d17-13d7-11e7-b5f5-060a60429fa6", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node <node> status is now: NodeNotReady
event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"<node>", UID:"ff9ba08d-13d7-11e7-b5f5-060a60429fa6", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node <node> status is now: NodeNotReady
event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"<node>", UID:"dec564bd-13d7-11e7-b5f5-060a60429fa6", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node <node> status is now: NodeNotReady
event.go:217] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"<node>", UID:"f8a909b7-13d7-11e7-b5f5-060a60429fa6", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeNotReady' Node <node> status is now: NodeNotReady
nodecontroller.go:608] NodeController detected that zone us-east-1::us-east-1b is now in state FullDisruption.
servicecontroller.go:640] Detected change in list of current cluster nodes. New node set: [<list of nodes that were not affected>]
servicecontroller.go:648] Successfully updated 53 out of 53 load balancers to direct traffic to the updated set of nodes
```

I did temporarily break access to my API server to see if that would trigger it, but it did not.

Any suggestions to additional logging I should enable on the kubelet? It does seem like a rare occurrence so I'm hoping I can capture everything possible for the next time it pops up.

**Comment by jswoods** [comment]
Just to follow up here: after more investigation, I believe we hit the issue explained in https://github.com/kubernetes/kubernetes/issues/30972#issuecomment-241077740. In attempt to avoid this scenario in the future, we're attempting to resolve any API server slowness by increasing the master instance sizes and also setting the following flags on our controller manager:
```
--node-monitor-grace-period=300s \
--node-eviction-rate=.00333 \
--secondary-node-eviction-rate=.00333
```

**Comment by jswoods** [comment]
Okay, one more update. I'm fairly confident that this issue is related to having the API servers behind an ELB, and having kubelet communication to the API servers go through the ELB. It does appear this issue shows up when the ELB addresses change (usually during a scale up/down of the ELB, e.g. during or following a spike in API traffic). The TTL on the ELB aliases are 60 seconds, and the default node monitor grace period is 40 seconds. This gives a window where the ELB address could change on the kubelets which causes them to temporarily be unable to report their health, but the controller manager marks the nodes as down (after 40 seconds) and starts evicting pods before the kubelets see the DNS update and are able to report a healthy status back to the API server. It's my suggestion that if you're running the API servers behind an ELB, then the `--node-monitor-grace-period` should be set to 60 seconds or more. The other options such as `--node-eviction-rate` and `--secondary-node-eviction-rate` will also reduce the aggressiveness of the pod eviction which could be useful if the communication from the kubelet to the API server is severed for a longer period of time. Note that I'm currently running the above configuration.

**Comment by jsravn** [comment]
@jswoods I just had the same thing happen - NodeNotReady for about 15 minutes. No errors in kubelet logs - which means kubelet thought it successfully posted the node status (unless there is a bug in this logic). I'm also using HA masters behind an ELB.

Are you sure it's an ELB problem? Because I would expect to see kubelet report an error in that case. Note I've also seen some other strange things which I believe are due to the ELB - e.g. streamwatcher.go logs an error occasionally about the watch stream failing, but I think this is a separate issue.

I'm thinking the more likely explanation is not using `--etcd-quorum-read` which means the controller manager could get stale reads despite kubelet successfully updating itself.

I'm going to dig a bit more into the kubelet code and see if it is possible for it to run into an error with the ELB and not actually report any errors (like maybe a really long/no timeout?).

**Comment by jsravn** [comment]
Also to add - we long ago set those eviction rates to be really low, because we very early on discovered what happened when there was a network glitch, or apiserver down temporarily (DELETE THE WORLD!) :). 1.5 is finally getting sensible and has removed the eviction altogether, 1.6 lets you specific eviction policy at a pod level. On 1.4 and prior, everyone should set those values correctly to be safe.

**Comment by jsravn** [comment]
It's really unfortunate that NodeNotReady event doesn't include the reason. It seems like the best you can do is use --v=4 in your controller manager - that looks like it will log most of the reasons in the node controller, but it might be too noisy for normal use.

**Comment by jswoods** [comment]
@jsravn From my testing, I could reproduce this issue by either deleting the ELB or breaking the access to the api server from the kubelet in another way (e.g. making a bad entry in /etc/hosts for the kubernetes master hostname and restarting the kubelet). I confirmed the kubelet does not log that it cannot send its status to the api server.

Interesting point about the `--etcd-quorum-read` flag. I will look into that a bit.

Also what led to the ELB being the issue is that while the kubelet doesn't log that it can't send its status, it did a few minutes later log `trypullsecret: read tcp <ip>:38600-><ip>:443: read: connection reset by peer`. Where the IP was one of the ELB's IPs. Note that it logged that message right around the same time the ELB had a change in IP addresses, which is how we figured this was being triggered when the ELB's IPs change and the kubelet has temporary stale DNS since the TTL for the dns ELB record is 60 seconds, while the amount of time the controller manager will wait for a kubelet to send a status before declaring it dead is only 40 seconds. We have not hit this issue again since setting `--node-monitor-grace-period` to 5 minutes.

**Comment by jsravn** [comment]
@jswoods how do you monitor when your ELB IPs change? I don't think it's the read quorum thing, the amount of time is too long.

**Comment by jswoods** [comment]
@jsravn We just noticed that the IP address it was connecting to and getting the `read: connection reset by peer` error was a different IP than what it was at the current point in time, so we assume it changed. According to AWS, the ELB IPs change from time to time and there is nothing you can do about it. Our solution was rather than monitoring for changing IPs, we set the monitor grace period threshold to be greater than the TTL for the ELB DNS record. That way even when the ELB IP address changes, the kubelet has 5 minutes to get the updated DNS record before the controller considers it unhealthy.

**Comment by jsravn** [comment]
@jswoods I think you're right. In my VPC flow logs I found what looks like an apiserver ELB network interface that stopped 50s before my node went not ready. I didn't get any errors at all in my kubelet logs though, so I wonder if there is some transparent retry mechanism with very long timeout that is getting hit. What a pain - now I really need to look into something to replace the ELB.

**Comment by jsravn** [comment]
Yes the ELB resizes itself and changes IPs when that happens. It's one of the worst things about ELBs, because it's very disruptive to all traffic (it abruptly closes all connections at the end of the 60s window). But it happens infrequently enough that we put up with it. I'm not sure having a longer grace period will help, because kubelet maintains persistent connections to the ELB which it will not know got closed. In my case I observed it took 15 minutes for kubelet to recover, so 5m would have not helped me :(.

**Comment by jsravn** [comment]
If only kube-proxy supported multiple apiserver addresses :). I think I'll need to put together some utility that registers each node in the ASG with a route53 entry on startup, then I can just use DNS load balancing.

**Comment by jswoods** [comment]
@jsravn interesting point about the persistent connections. For what it's worth, we have not seen this issue since making the adjustment to the grace period. And the problem happened every day for the previous 3 days before making the change. I'll post an update though if anything changes.

**Comment by tarvip** [comment]
> I think I'll need to put together some utility that registers each node in the ASG with a route53 entry on startup, then I can just use DNS load balancing.

@jsravn You could use something like this: https://objectpartners.com/2015/07/07/aws-tricks-updating-route53-dns-for-autoscalinggroup-using-lambda/
I have tested this approach (although not with kubernetes) and it seems to work fine.

**Comment by glesage** [comment]
I'm seeing this exact issue currently on Google Cloud Container Engine with k8s 1.6.0

We're also just scaling up the size of the nodes in order to offset this issue as @jswoods mentioned, but this is pretty expensive for us (we're a startup).

I've found a LOT of other issues that seem related, or essentially the same thing on other providers, or other `ideas` for what the issue is. Would love some more insight on how I can provide information on this, logs galore!

https://github.com/kubernetes/kubernetes/issues/41904
https://github.com/kubernetes/kubernetes/issues/43516
https://github.com/kubernetes/kubernetes/issues/44458
https://github.com/kubernetes/kubernetes/issues/41850
https://github.com/kubernetes/kubernetes/issues/32522
https://github.com/kubernetes/kubernetes/issues/31415
https://github.com/kubernetes/kubernetes/issues/30972
https://github.com/kubernetes/kubernetes/issues/26566

Seems like these are the origins posts from a while back, may no longer be relevant since this was for k8s 1.1
https://github.com/kubernetes/kubernetes/issues/24200
https://github.com/kubernetes/kubernetes/issues/19124
https://github.com/kubernetes/kubernetes/issues/13412
https://github.com/kubernetes/kubernetes/issues/13398

I'll see if I can help with writting these: https://github.com/kubernetes/kubernetes/issues/13429

**Comment by kargakis** [comment]
@kubernetes/sig-node-bugs

**Comment by dchen1107** [comment]
@glesage Could you please open a new issue related to Google Cloud Container Engine with k8s 1.6.0? You can directly assign it to me. 

The reason of NodeNotReady vary in different production environment. And this issue was filed initially for AWS Kubernetes, and the rest discussion is also focusing on that.

**Comment by dchen1107** [comment]
@zmerlynn I am re-assigning this to you since you have more context on AWS than me. Feel free to re-assign back to sig-node team or myself once we ruled out the cloud provider / configuration issues. Thanks!

**Comment by glesage** [comment]
I see, no problem will do thks @dchen1107

**Comment by Raffo** [comment]
I've seen a similar issue, but I'm not sure if it's related. I restart the apiserver by changing the manifest on the server nodes. There was a time while the apiserver was not running and the node transitioned to the `NotReady` state. After that all the daemonsets that we run on the master nodes became `Terminating` and I could verify they were not running on the master nodes.

**Comment by jswoods** [comment]
We've seen this issue again even after setting the flags as mentioned earlier. We're now looking for alternative ways to deploy H/A masters. Are there any suggestions on how to do that while avoiding a load balancer?

**Comment by willmore** [comment]
>Are there any suggestions on how to do that while avoiding a load balancer?

I believe one solution is to use DNS based routing to the masters (round-robin resolution for example).

**Comment by jsravn** [comment]
I've swapped to DNS, our problems seem to have been solved so far. Will update if I see an issue again.

**Comment by glesage** [comment]
For any whom this might help; I we ended up getting it to work by optimizing our pods to take a whole lot less memory. Not sure if this is related to you guys' issues but our app was taking over 1 gig of memory at peak... getting that down to 200MB or less fixed the issue for us.

The Node the pods resided on had plenty of memory to go though so I'm still curious to see what was happening but at least our service isn't going down anymore

Thanks!

**Comment by zihaoyu** [comment]
We hit the same issue twice in our production cluster in aws us-west-2 vpc. The last occurrence was March 8. Each time it caused a brief period of disruption to the cluster since some pods of production services were rescheduled. 

On March 8, from `kubelet` logs we saw a lot of messages like the following:

```
E0308 02:02:09.221567   43003 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/secret/1bf35246-f86f-11e6-8025-028c35746a57-default-token-hwncv\" (\"1bf35246-f86f-11e6-8025-028c35746a57\")" failed. No retries permitted until 2017-03-08 02:02:09.721554338 +0000 UTC (durationBeforeRetry 500ms). Error: MountVolume.SetUp failed for volume "kubernetes.io/secret/1bf35246-f86f-11e6-8025-028c35746a57-default-token-hwncv" (spec.Name: "default-token-hwncv") pod "1bf35246-f86f-11e6-8025-028c35746a57" (UID: "1bf35246-f86f-11e6-8025-028c35746a57") with: Get https://k8s-master-elb.mycompany.com/api/v1/namespaces/production/secrets/default-token-hwncv: read tcp 10.72.33.48:53066->10.72.2.91:443: read: connection timed out
```

Note the ELB IP `10.72.2.91` that `kubelet` was trying to connect to. We then opened tickets with AWS. They confirmed that: (below is the response from AWS support)

> During a natural ELB scaling event, IP addresses 10.72.2.91 and 10.72.xx.xxx were confirmed to be removed from this ELB's DNS at 2017-03-02T00:47:52Z UTC (March 2nd). Clients should not have been communicating with it after this point if respecting a DNS TTL of 60 seconds.

> We kept the nodes around for quite a while between March 2nd and March 8th as they were evidently still receiving traffic from clients. These nodes with these IP addresses were then removed from the ELB at 2017-03-08T01:46:34Z (March 8th).

> Per the sample logs provided, your application sitting in front of this ELB was trying to communicate with the old node IP addresses @ 2017-03-08T02:02:09 / 2017-03-08T02:02:13 UTC, long after that IP was removed from DNS during natural scaling.

> there definitely appears to be a similar strange caching behavior on the clients side where it is not respecting a DNS TTL of 60 seconds.

> After the IPs were removed from the ELB, any DNS query to the ELB DNS name will no longer return the IPs. The node usually stay around for a while to allow request already sent to it to complete, there is no fixed time for how long the nodes will be ariound, it could range betweeen couple of hours to days. In your case, the nodes were not removed for days and your application still kept sending requests to the nodes. I believe you started facing this issue after the nodes were eventually removed.

> The solution has my colleagues has already mentioned is to honor the TTL valus of the ELB to avaoid such issues.


I suspect `kubelet` uses some kind of keep-alive connections to the `apiserver`, and it does not honor TTL. Is there such possibility?

**Comment by jsravn** [comment]
Of course, kubelet uses http/1.1. With long polling kubelet keeps the connection open for a long time. I always regarded this behaviour of ELBs breaking all connections at scale up/down to be broken. It doesnt just affect kubelet, it's any traffic going through an ELB. They should at least honour configured idle times on the ELB but they don't - it's hard coded to 60s. I'm not quite sure though why kubelet doesn't detect the connection loss sooner - I suspect the ELBs are not properly sending a FIN. AWS really should be doing something like using keepalived and using virtual IPs to seamlessly transition instances without affecting customers. But they don't. Fortunately scale events are pretty rare or this would make ELBs generally unusable for anyone (and for those that matter, you can actually ask aws to freeze the scaling, and even prescale for you, but it's a painful process and they don't like to do it).

**Comment by redbaron** [comment]
> If only kube-proxy supported multiple apiserver addresses :). I think I'll need to put together some utility that registers each node in the ASG with a route53 entry on startup, then I can just use DNS load balancing.

@jsravn , I am trying to understand, how it is different from ELB changing IP addresses? You roll your controllers ASG, DNS record changes, kubelets which started before you ASG roll continue to use old Apiservers which are down by that time. I see below you say ELB-less DNS-only approach helped you, can you confirm that you can trigger controller nodes ASG roll and things continue to work please?

**Comment by jsravn** [comment]
@redbaron My ASG IPs don't change. Even if they did, I could shut down apiserver before taking it out, this would let kubelet properly disconnect. For rolling update of ASG, I shutdown apiserver as well which lets kubelet properly disconnect and then try to reconnect.

**Comment by eveny** [comment]
I see similar behaviour.  The physical cluster with 3 nodes, multi-master setup.
**Environment:**
K8S: 1.6.2
OS: Container Linux by CoreOS 1353.6.0 (Ladybug)
Kernel: 4.9.24-coreos
h1: 144.76.41.106
h2: 144.76.62.71
h3: 136.243.71.6
load balancer: 148.251.206.80


**Steps to reproduce:**

1. Stable cluster, Load balancer IP is on h2 and it behaves as master.

Kubernetes master is running at https://148.251.206.80:6443

**kubectl get nodes**
```
NAME      STATUS    AGE
h1        Ready     21h
h2        Ready     21h
h3        Ready     21h
```

**kubectl get pods --all-namespaces -o wide**
```
NAMESPACE     NAME                                    READY     STATUS    RESTARTS   AGE       IP              NODE
default       weave-scope-app-681980193-dg4bq         1/1       Running   0          1h        10.100.96.1     h3
kube-system   heapster-2315332064-jf69p               1/1       Running   0          1h        10.100.96.2     h3
kube-system   kube-apiserver-h1                       1/1       Running   0          1h        144.76.41.106   h1
kube-system   kube-apiserver-h2                       1/1       Running   1          1h        144.76.62.71    h2
kube-system   kube-apiserver-h3                       1/1       Running   0          1h        136.243.71.6    h3
kube-system   kube-controller-manager-h1              1/1       Running   0          1h        144.76.41.106   h1
kube-system   kube-controller-manager-h2              1/1       Running   2          1h        144.76.62.71    h2
kube-system   kube-controller-manager-h3              1/1       Running   1          1h        136.243.71.6    h3
kube-system   kube-dns-2089938067-fh2d5               3/3       Running   0          1h        10.100.96.3     h3
kube-system   kube-dns-2089938067-x6k54               3/3       Running   0          1h        10.100.0.2      h2
kube-system   kube-dns-autoscaler-1866472577-600bd    1/1       Running   0          1h        10.100.96.4     h3
kube-system   kube-proxy-h1                           1/1       Running   0          1h        144.76.41.106   h1
kube-system   kube-proxy-h2                           1/1       Running   3          1h        144.76.62.71    h2
kube-system   kube-proxy-h3                           1/1       Running   0          1h        136.243.71.6    h3
kube-system   kube-scheduler-h1                       1/1       Running   0          1h        144.76.41.106   h1
kube-system   kube-scheduler-h2                       1/1       Running   3          1h        144.76.62.71    h2
kube-system   kube-scheduler-h3                       1/1       Running   1          1h        136.243.71.6    h3
kube-system   kubernetes-dashboard-2396447444-gcrzk   1/1       Running   0          1h        10.100.96.5     h3
kube-system   monitoring-grafana-3975459543-b4t8x     1/1       Running   0          1h        10.100.0.3      h2
kube-system   monitoring-influxdb-3480804314-vb717    1/1       Running   0          1h        10.100.96.6     h3
kube-system   prometheus-operator-3001283097-3ctt9    1/1       Running   0          1h        10.100.0.5      h2
kube-system   weave-net-5l0t7                         2/2       Running   0          21h       136.243.71.6    h3
kube-system   weave-net-6l5r1                         2/2       Running   0          21h       144.76.41.106   h1
kube-system   weave-net-x0sb7                         2/2       Running   2          21h       144.76.62.71    h2
```

**established connections on h1**
```
core@h1 ~ $ sudo netstat -tuepn | grep 148.251.206.80
tcp        0      0 144.76.41.106:42616     148.251.206.80:6443     ESTABLISHED 0          1083225    29862/hyperkube
tcp        0      0 144.76.41.106:45240     148.251.206.80:6443     ESTABLISHED 0          1103586    29257/kubelet

core@h1 ~ $ ps -ef | grep 29862
root     29862 29845  0 Apr26 ?        00:01:47 /hyperkube proxy --master=https://148.251.206.80:6443 --bind-address=144.76.41.106 --proxy-mode=iptables --masquerade-all=true --cluster-cidr=10.100.0.0/16 --hostname-override=h1 --kubeconfig=/etc/kubernetes/configs/proxy.yml --healthz-bind-address=127.0.0.1 --healthz-port=10249

core@h1 ~ $ ps -ef | grep 29257
root     29257     1  2 Apr26 ?        00:31:32 /kubelet --address=144.76.41.106 --port=10250 --require-kubeconfig --kubeconfig=/etc/kubernetes/configs/kubelet.yml --allow-privileged=true --container-runtime=docker --docker-endpoint=unix:///var/run/docker.sock --register-node=true --network-plugin=cni --network-plugin-dir=/etc/cni/net.d --pod-manifest-path=/etc/kubernetes/manifests --hostname-override=h1 --cluster-dns=10.3.0.10 --cluster-domain=fns.local --tls-cert-file=/etc/kubernetes/pki/controller.crt --tls-private-key-file=/etc/kubernetes/pki/controller.key --node-status-update-frequency=5s
```

**established connections on h2**
```
core@h2 ~ $ sudo netstat -tuepn | grep 148.251.206.80
tcp        0      0 148.251.206.80:38646    148.251.206.80:6443     ESTABLISHED 0          1295207    2659/hyperkube
tcp        0      0 148.251.206.80:38588    148.251.206.80:6443     ESTABLISHED 0          1295155    1471/kubelet
tcp6       0      0 148.251.206.80:6443     136.243.71.6:47616      ESTABLISHED 0          1303620    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:45240     ESTABLISHED 0          1335414    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     10.100.0.2:48490        ESTABLISHED 0          1302595    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:38488      ESTABLISHED 0          1305719    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     10.100.0.5:38938        ESTABLISHED 0          1302823    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:34682      ESTABLISHED 0          1303579    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     148.251.206.80:38588    ESTABLISHED 0          1294194    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:47200      ESTABLISHED 0          1297690    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:48326      ESTABLISHED 0          1298045    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:58590      ESTABLISHED 0          1299301    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:34134     ESTABLISHED 0          1294169    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.62.71:33156      ESTABLISHED 0          1308911    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:42616     ESTABLISHED 0          1306556    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:48232      ESTABLISHED 0          1295927    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     148.251.206.80:38646    ESTABLISHED 0          1298685    2114/hyperkube

core@h2 ~ $ ps -ef | grep 1471
root      1471     1  2 Apr26 ?        00:32:27 /kubelet --address=144.76.62.71 --port=10250 --require-kubeconfig --kubeconfig=/etc/kubernetes/configs/kubelet.yml --allow-privileged=true --container-runtime=docker --docker-endpoint=unix:///var/run/docker.sock --register-node=true --network-plugin=cni --network-plugin-dir=/etc/cni/net.d --pod-manifest-path=/etc/kubernetes/manifests --hostname-override=h2 --cluster-dns=10.3.0.10 --cluster-domain=fns.local --tls-cert-file=/etc/kubernetes/pki/controller.crt --tls-private-key-file=/etc/kubernetes/pki/controller.key --node-status-update-frequency=5s


core@h2 ~ $ ps -ef | grep 2659
root      2659  2643  0 Apr26 ?        00:01:46 /hyperkube proxy --master=https://148.251.206.80:6443 --bind-address=144.76.62.71 --proxy-mode=iptables --masquerade-all=true --cluster-cidr=10.100.0.0/16 --hostname-override=h2 --kubeconfig=/etc/kubernetes/configs/proxy.yml --healthz-bind-address=127.0.0.1 --healthz-port=10249


core@h2 ~ $ ps -ef | grep 2114
root      2114  2097  0 Apr26 ?        00:10:09 /hyperkube apiserver --bind-address=0.0.0.0 --secure-port=6443 --insecure-bind-address=127.0.0.1 --insecure-port=8080 --etcd-servers=https://127.0.0.1:2379 --etcd-cafile=/etc/etcd/pki/etcd-k8s-ca.crt --etcd-certfile=/etc/etcd/pki/etcd-k8s.crt --etcd-keyfile=/etc/etcd/pki/etcd-k8s.key --allow-privileged=true --service-cluster-ip-range=10.3.0.0/16 --advertise-address=148.251.206.80 --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota --tls-cert-file=/etc/kubernetes/pki/controller.crt --tls-private-key-file=/etc/kubernetes/pki/controller.key --client-ca-file=/etc/kubernetes/pki/etcd-k8s-ca.crt --runtime-config=extensions/v1beta1/networkpolicies=true --anonymous-auth=false --storage-backend=etcd3 --kubelet-https=true --kubelet-preferred-address-types=InternelIP,ExternalIP,Hostname --kubelet-certificate-authority=/etc/kubernetes/pki/etcd-k8s-ca.crt --kubelet-client-certificate=/etc/kubernetes/pki/controller.crt --kubelet-client-key=/etc/kubernetes/pki/controller.key --requestheader-username-headers=X-Remote-User --requestheader-group-headers=X-Remote-Group --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-allowed-names=front-proxy-client --requestheader-client-ca-file=/etc/kubernetes/pki/etcd-k8s-ca.crt --authorization-mode=AlwaysAllow --apiserver-count=3 --basic-auth-file=/etc/kubernetes/accounts.csv
```
**established connections on h3**
```
core@h3 ~ $ sudo netstat -tuepn | grep 148.251.206.80
tcp        0      0 136.243.71.6:48232      148.251.206.80:6443     ESTABLISHED 0          1281724    29852/hyperkube
tcp        0      0 136.243.71.6:48326      148.251.206.80:6443     ESTABLISHED 0          1284536    29303/kubelet

core@h3 ~ $ ps -ef | grep 29852
root     29852 29835  0 Apr26 ?        00:01:58 /hyperkube proxy --master=https://148.251.206.80:6443 --bind-address=136.243.71.6 --proxy-mode=iptables --masquerade-all=true --cluster-cidr=10.100.0.0/16 --hostname-override=h3 --kubeconfig=/etc/kubernetes/configs/proxy.yml --healthz-bind-address=127.0.0.1 --healthz-port=10249

core@h3 ~ $ ps -ef | grep 29303
root     29303     1  2 Apr26 ?        00:31:47 /kubelet --address=136.243.71.6 --port=10250 --require-kubeconfig --kubeconfig=/etc/kubernetes/configs/kubelet.yml --allow-privileged=true --container-runtime=docker --docker-endpoint=unix:///var/run/docker.sock --register-node=true --network-plugin=cni --network-plugin-dir=/etc/cni/net.d --pod-manifest-path=/etc/kubernetes/manifests --hostname-override=h3 --cluster-dns=10.3.0.10 --cluster-domain=fns.local --tls-cert-file=/etc/kubernetes/pki/controller.crt --tls-private-key-file=/etc/kubernetes/pki/controller.key --node-status-update-frequency=5s
```

2. Migrate load balancer to h3.

Kubernetes master is running at https://148.251.206.80:6443

**kubectl get nodes**
```
NAME      STATUS     AGE
h1        Ready      21h
h2        NotReady   21h
h3        Ready      21h
```

**kubectl get pods --all-namespaces -o wide**
```
NAMESPACE     NAME                                    READY     STATUS     RESTARTS   AGE       IP              NODE
default       weave-scope-app-681980193-63s65         1/1       Running    0          1m        10.100.128.2    h1
default       weave-scope-app-681980193-wmww7         1/1       Unknown    0          2m        10.100.0.4      h2
kube-system   heapster-2315332064-kd2vw               1/1       Unknown    0          2m        10.100.0.6      h2
kube-system   heapster-2315332064-z8wlm               1/1       Running    0          1m        10.100.128.1    h1
kube-system   kube-apiserver-h1                       1/1       Running    0          1m        144.76.41.106   h1
kube-system   kube-apiserver-h2                       1/1       Unknown    1          1h        144.76.62.71    h2
kube-system   kube-apiserver-h3                       1/1       Running    0          1m        136.243.71.6    h3
kube-system   kube-controller-manager-h1              1/1       Running    0          1m        144.76.41.106   h1
kube-system   kube-controller-manager-h2              1/1       Unknown    2          1h        144.76.62.71    h2
kube-system   kube-controller-manager-h3              1/1       Running    1          1m        136.243.71.6    h3
kube-system   kube-dns-2089938067-9l6hp               3/3       Running    0          1m        10.100.96.1     h3
kube-system   kube-dns-2089938067-qxjlv               3/3       Running    0          1m        10.100.128.3    h1
kube-system   kube-dns-2089938067-t7dhs               3/3       Unknown    0          2m        10.100.0.7      h2
kube-system   kube-dns-2089938067-x6k54               3/3       Unknown    0          1h        10.100.0.2      h2
kube-system   kube-dns-autoscaler-1866472577-2lk9x    1/1       Running    0          1m        10.100.128.4    h1
kube-system   kube-dns-autoscaler-1866472577-xgvtz    1/1       Unknown    0          2m        10.100.0.8      h2
kube-system   kube-proxy-h1                           1/1       Running    0          1m        144.76.41.106   h1
kube-system   kube-proxy-h2                           1/1       Unknown    3          1h        144.76.62.71    h2
kube-system   kube-proxy-h3                           1/1       Running    0          1m        136.243.71.6    h3
kube-system   kube-scheduler-h1                       1/1       Running    0          1m        144.76.41.106   h1
kube-system   kube-scheduler-h2                       1/1       Unknown    3          1h        144.76.62.71    h2
kube-system   kube-scheduler-h3                       1/1       Running    1          1m        136.243.71.6    h3
kube-system   kubernetes-dashboard-2396447444-54rft   1/1       Unknown    0          2m        10.100.0.9      h2
kube-system   kubernetes-dashboard-2396447444-67gt1   1/1       Running    0          1m        10.100.128.5    h1
kube-system   monitoring-grafana-3975459543-b4t8x     1/1       Unknown    0          1h        10.100.0.3      h2
kube-system   monitoring-grafana-3975459543-tnl1x     1/1       Running    0          1m        10.100.96.2     h3
kube-system   monitoring-influxdb-3480804314-b6h3m    1/1       Unknown    0          2m        10.100.0.10     h2
kube-system   monitoring-influxdb-3480804314-bxt66    1/1       Running    0          1m        10.100.128.6    h1
kube-system   prometheus-operator-3001283097-1pbvv    1/1       Running    0          1m        10.100.96.4     h3
kube-system   prometheus-operator-3001283097-3ctt9    1/1       Unknown    0          1h        10.100.0.5      h2
kube-system   weave-net-5l0t7                         2/2       Running    0          21h       136.243.71.6    h3
kube-system   weave-net-6l5r1                         2/2       Running    0          21h       144.76.41.106   h1
kube-system   weave-net-x0sb7                         2/2       NodeLost   2          21h       144.76.62.71    h2
```

**established connections on h1**
```
core@h1 ~ $ sudo netstat -tuepn | grep 148.251.206.80
tcp        0      0 144.76.41.106:50886     148.251.206.80:6443     ESTABLISHED 0          1141699    29257/kubelet
tcp        0      0 144.76.41.106:50832     148.251.206.80:6443     ESTABLISHED 0          1142482    29862/hyperkube
```

**established connections on h2**
```
core@h2 ~ $ sudo netstat -tuepn | grep 148.251.206.80
tcp        0      0 148.251.206.80:38646    148.251.206.80:6443     ESTABLISHED 0          1295207    2659/hyperkube
tcp        0    315 148.251.206.80:38588    148.251.206.80:6443     ESTABLISHED 0          1295155    1471/kubelet
tcp6       0      0 148.251.206.80:6443     136.243.71.6:47616      ESTABLISHED 0          1303620    2114/hyperkube
tcp6       0  36579 148.251.206.80:6443     144.76.41.106:45240     ESTABLISHED 0          1335414    2114/hyperkube
tcp6       0  47432 148.251.206.80:6443     10.100.0.2:48490        ESTABLISHED 0          1302595    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:38488      ESTABLISHED 0          1305719    2114/hyperkube
tcp6       0     76 148.251.206.80:6443     10.100.0.5:38938        ESTABLISHED 0          1302823    2114/hyperkube
tcp6       0  41931 148.251.206.80:6443     136.243.71.6:34682      ESTABLISHED 0          1303579    2114/hyperkube
tcp6       0  98995 148.251.206.80:6443     148.251.206.80:38588    ESTABLISHED 0          1294194    2114/hyperkube
tcp6       0    114 148.251.206.80:6443     136.243.71.6:47200      ESTABLISHED 0          1297690    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     10.100.0.8:34780        ESTABLISHED 0          1453866    2114/hyperkube
tcp6       0  49615 148.251.206.80:6443     136.243.71.6:48326      ESTABLISHED 0          1298045    2114/hyperkube
tcp6       0 123971 148.251.206.80:6443     136.243.71.6:58590      ESTABLISHED 0          1299301    2114/hyperkube
tcp6       0  41885 148.251.206.80:6443     10.100.0.7:60394        ESTABLISHED 0          1453836    2114/hyperkube
tcp6       0  50312 148.251.206.80:6443     10.100.0.6:42710        ESTABLISHED 0          1456345    2114/hyperkube
tcp6       0      0 148.251.206.80:6443     10.100.0.9:55164        ESTABLISHED 0          1458655    2114/hyperkube
tcp6       0  78048 148.251.206.80:6443     144.76.41.106:34134     ESTABLISHED 0          1294169    2114/hyperkube
tcp6       0  53072 148.251.206.80:6443     144.76.62.71:33156      ESTABLISHED 0          1308911    2114/hyperkube
tcp6       0  37359 148.251.206.80:6443     144.76.41.106:42616     ESTABLISHED 0          1306556    2114/hyperkube
tcp6       0  41893 148.251.206.80:6443     136.243.71.6:48232      ESTABLISHED 0          1295927    2114/hyperkube
tcp6       0  75146 148.251.206.80:6443     148.251.206.80:38646    ESTABLISHED 0          1298685    2114/hyperkube
```

**established connections on h3**
```
core@h3 ~ $ sudo netstat -tuepn | grep 148.251.206.80
tcp        0      0 148.251.206.80:54274    148.251.206.80:6443     ESTABLISHED 0          1446588    29303/kubelet
tcp        0      0 148.251.206.80:54200    148.251.206.80:6443     ESTABLISHED 0          1444685    29852/hyperkube
tcp6       0      0 148.251.206.80:6443     148.251.206.80:54274    ESTABLISHED 0          1444829    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:50886     ESTABLISHED 0          1447607    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     10.100.96.1:50364       ESTABLISHED 0          1454085    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:43266     ESTABLISHED 0          1444645    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     148.251.206.80:54200    ESTABLISHED 0          1447458    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:34414     ESTABLISHED 0          1452227    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:50832     ESTABLISHED 0          1447416    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.62.71:47376      ESTABLISHED 0          1444797    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:40168     ESTABLISHED 0          1457332    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:39070     ESTABLISHED 0          1454367    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     10.100.96.4:35040       ESTABLISHED 0          1455402    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     144.76.41.106:55898     ESTABLISHED 0          1448940    29929/hyperkube
tcp6       0      0 148.251.206.80:6443     136.243.71.6:35072      ESTABLISHED 0          1458970    29929/hyperkube

core@h3 ~ $ ps -ef | grep 29929
root     29929 29912  1 Apr26 ?        00:14:10 /hyperkube apiserver --bind-address=0.0.0.0 --secure-port=6443 --insecure-bind-address=127.0.0.1 --insecure-port=8080 --etcd-servers=https://127.0.0.1:2379 --etcd-cafile=/etc/etcd/pki/etcd-k8s-ca.crt --etcd-certfile=/etc/etcd/pki/etcd-k8s.crt --etcd-keyfile=/etc/etcd/pki/etcd-k8s.key --allow-privileged=true --service-cluster-ip-range=10.3.0.0/16 --advertise-address=148.251.206.80 --admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota --tls-cert-file=/etc/kubernetes/pki/controller.crt --tls-private-key-file=/etc/kubernetes/pki/controller.key --client-ca-file=/etc/kubernetes/pki/etcd-k8s-ca.crt --runtime-config=extensions/v1beta1/networkpolicies=true --anonymous-auth=false --storage-backend=etcd3 --kubelet-https=true --kubelet-preferred-address-types=InternelIP,ExternalIP,Hostname --kubelet-certificate-authority=/etc/kubernetes/pki/etcd-k8s-ca.crt --kubelet-client-certificate=/etc/kubernetes/pki/controller.crt --kubelet-client-key=/etc/kubernetes/pki/controller.key --requestheader-username-headers=X-Remote-User --requestheader-group-headers=X-Remote-Group --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-allowed-names=front-proxy-client --requestheader-client-ca-file=/etc/kubernetes/pki/etcd-k8s-ca.crt --authorization-mode=AlwaysAllow --apiserver-count=3 --basic-auth-file=/etc/kubernetes/accounts.csv
```

Then I do connections reset for proxy, apiserver and kubelet on h2 in different combination and I see the problem is in connection between kubelet and api server, it is looked like it is long pooled with 15 minutes.
```
tcp        0      0 148.251.206.80:38588    148.251.206.80:6443     ESTABLISHED 0          1295155    1471/kubelet

core@h2 ~ $ ps -ef | grep 1471
root      1471     1  2 Apr26 ?        00:32:27 /kubelet --address=144.76.62.71 --port=10250 --require-kubeconfig --kubeconfig=/etc/kubernetes/configs/kubelet.yml --allow-privileged=true --container-runtime=docker --docker-endpoint=unix:///var/run/docker.sock --register-node=true --network-plugin=cni --network-plugin-dir=/etc/cni/net.d --pod-manifest-path=/etc/kubernetes/manifests --hostname-override=h2 --cluster-dns=10.3.0.10 --cluster-domain=fns.local --tls-cert-file=/etc/kubernetes/pki/controller.crt --tls-private-key-file=/etc/kubernetes/pki/controller.key --node-status-update-frequency=5s
```

If I do nothing, 15 minutes after load balancer ip switch between hosts this connection would be automatically closed and cluster would be in stable state.

If I delete apiserver and proxy pods on h2 and restart kubelet, issue is fixed. “Unknown” status form pods is changes to “Terminating”.

I still cannot understand behaviour when do only kubelet restart (I saw messages about orphaned pods, reboot fixed these messages).

How to manage "15 minutes” lag with configuration parameters?

**Comment by danielfm** [comment]
@eveny thanks for the information! 🎉 

Since you already have a test environment, can you do a quick test and use an ALB instead of a classic ELB, to see if this changes anything?

**Comment by danielfm** [comment]
Already linked my kube-aws issue to this one, but just for the record I'd like to report that I've run into this issue twice in 3 days, and in both occasions ~50% of the nodes transitioned to `NotReady` at roughly the same time causing ~10m downtime in some critical services running in my production cluster.

Just like @zihaoyu reported, I also contacted the AWS support requesting information about any IP changes in the controller ELB, and they confirmed ELB nodes were shut down at both occasions.

**Comment by eveny** [comment]
@danielfm. I was not clear. I tested a setup on 3 bare metal servers on hetzner. Maybe this thread is not a best place to describe the issue but there are a bunch of tickets with similar behaviour which are linked together.
Actually I don't have AWS
Once again there is the setup.
```
Initial state:
                          ________________________
                         |                        |
                         | VIP: Unassigned        |
                   |-----| Host IP: 144.76.41.106 |
                   |     | Role: Slave            |
                   |     |________________________|
                   |
                   |      ________________________
                   |     |                       |
                   |     | VIP: 148.251.206.80   |
(148.251.206.80)---|-----| Host IP: 144.76.62.71 |
                   |     | Role: Master          |
                   |     |_______________________|
                   |
                   |      _______________________
                   |     |                       |
                   |     | VIP: Unassigned       |
                   |-----| Host IP: 136.243.71.6 |
                         | Role: Slave           |
                         |_______________________|

Switch:

                          ________________________
                         |                        |
                         | VIP: Unassigned        |
                   |-----| Host IP: 144.76.41.106 |
                   |     | Role: Slave            |
                   |     |________________________|
                   |
                   |      ________________________
                   |     |                       |
                   |     | VIP: Unassigned       |
(148.251.206.80)---|-----| Host IP: 144.76.62.71 |
                   |     | Role: Slave           |
                   |     |_______________________|
                   |
                   |      _______________________
                   |     |                       |
                   |     | VIP: 148.251.206.80   |
                   |-----| Host IP: 136.243.71.6 |
                         | Role: Master          |
                         |_______________________|
```
And I see the issue in configuration:
1. When I do config  **/kubelet --address=144.76.62.71** I assume that connections will be initiated from this IP address. I see established connections from floating IP and Send-Q is not empty when I do switch of IP to another host.
```
Proto Recv-Q Send-Q Local Address           Foreign Address         State       User       Inode      PID/Program name
tcp        0    315 148.251.206.80:38588    148.251.206.80:6443     ESTABLISHED 0          1295155    1471/kubelet
```
There are could be plenty of reasons to setup additional IPs on interface and making configuration like **--address=** should guarantee that cluster component uses this particular IP.

2. After switch kubelet waits too long before reestablish connection to api server (it looks like 15 mins). This behaviour should be configurable with parameter.

**Comment by jsravn** [comment]
@eveny `--address` is for kubelet's http server. It's not going to affect the originating IP to apiserver, which I suspect will simply be the address of the outbound interface. 15m is probably the resync period built into the kube client.

**Comment by danielfm** [comment]
@jsravn perhaps [this](https://github.com/kubernetes/client-go/blob/65ac4e638a79d49b7dbd5e1cef4e7785af7cbd39/tools/cache/reflector.go#L78) is it?

**Comment by jsravn** [comment]
@danielfm I don't think kubelet uses the informer - and I think that just relists the cache, it doesn't actually list against the apiserver directly. I tried looking into it, and afaict there is no logic to relist periodically.

**Comment by jfoy** [comment]
We've observed this same failure behavior in AWS. Our architecture is as described above: kubelets connect to apiserver through an ELB discovered via Route 53 DNS.

The apiserver lost connections from the nodes coming through one ELB IP first, followed by those coming through the other IP about 25 seconds later; the nodes were marked NodeNotReady. The nodes were unable to reach the apiserver for 15 minutes, after which the kubelet TCP connections timed out, re-established connections to the apiserver through new ELB IPs, and the nodes returned to Ready. During this window the apiserver continued to respond to ELB health checks. 

Our AWS support staff tells us that the scaling event removing the IPs in question from ELB DNS occurred 7 days prior to the outage, and that they only executed a hard drop on connections to the old interfaces after that window. This suggests that kubelet persistent connections to the apiserver had been kept alive through that entire period.

I don't see any configuration options on kubelet for tuning HTTP keep-alive behavior, but it seems like this might be a natural first work-around. I'll file a separate issue calling out that requirement.

**Comment by jsravn** [comment]
Okay, another update.

Switching to DNS does *not* help for this issue. We just had another production issue with kube-proxy being an hour (!!!!) out of date, then finally getting the `streamwatcher.go:103] Unexpected EOF during watch stream event decoding: unexpected EOF` message before updating. This caused our endpoints to be out of date and requests were failing.

This is pretty bad, needless to say.

After some testing, it doesn't appear kube-apiserver terminates gracefully. That is, when I run the kube-apiserver docker container, I can't stop it. If you try to manually do docker stop (which sends SIGTERM), kube-apiserver will just chug along and ignore the SIGTERM until docker SIGKILLs it. As a result, all connections to apiserver will be stuck in a bad state.

If apiserver properly shut down itself, it should send FIN to all of its clients, which will then immediately close their connections and reconnect to another apiserver. I'm not sure what exactly is going on here but it's a serious bug IMO.

**Comment by redbaron** [comment]
> If apiserver properly shut down itself, it should send FIN to all of its clients
and if it is `SIGKILL`ed then Linux kernel will send RST packet on any incoming packet coming from kubelet, which would immediately trigger closed connection on kubelet side

**Comment by jsravn** [comment]
@redbaron That doesn't help with kubelet and kube-proxy, which send no data. They do a long poll and just sit blocked on a read. apiserver needs to send the FIN otherwise those connections just hang until kubelet/kube-proxy try to resync or whatever.

**Comment by jsravn** [comment]
It's possible that tuning your OS tcp keepalive settings might help mitigate this, but something is wrong w/ apiserver I think - or at least with the docker container of it.

**Comment by jsravn** [comment]
It also seems like the resync periods in kube-proxy/kubelet (which is 15m by default) is not working properly. Since in the most recent case I had, it took about 1 hour for kube-proxy to detect the apiserver connection was stale. I had a critical pod redeployed, and it took 26 mins for the endpoint to be updated (as seen in kube-proxy logs with `-v=2`). The apiserver instance was restarted an hour prior to the endpoint being removed.

**Comment by jswoods** [comment]
We've resorted to running this every 15 seconds as a workaround https://gist.github.com/jswoods/ae5fe375f206c1e1787139ebbdaa39f2. Not ideal at all but so far as kept our cluster up.

**Comment by jsravn** [comment]
On more thought, kubelet should *really* be checking the health of the watches. I had always thought resync is supposed to do that, but maybe that logic has changed (or my understanding is broken)? Relying on a long lived TCP connection to stay up in the presence of middleware, load balancing, and so on is not a great idea without application level pings. I'll have to dig more into the resync code.

**Comment by redbaron** [comment]
> On more thought, kubelet should really be checking the health of the watches
if it is long polling, then it is apiserver which should be sending noop messages, which are then to be ignored by kubelet

**Comment by jsravn** [comment]
I spent some time trying to reproduce this without much luck. Even if I kill -9 apiserver kube-proxy/kubelet seem to recovery pretty quickly, detect an EOF, and reconnect. The worst I could get was about a minute downtime by rebooting an apiserver node. So I'm not sure what is causing the behaviour where kubelet/kube-proxy seem to stop processing updates for a long time.

**Comment by jsravn** [comment]
kube-proxy/kubelet already enable keepalive_time of 30s with interval of 30s. This is strace on socket options when it reconnects:
```
[pid  2310] setsockopt(6, SOL_TCP, TCP_NODELAY, [1], 4) = 0
[pid  2310] setsockopt(6, SOL_SOCKET, SO_KEEPALIVE, [1], 4) = 0
[pid  2310] setsockopt(6, SOL_TCP, TCP_KEEPINTVL, [30], 4) = 0
[pid  2310] setsockopt(6, SOL_TCP, TCP_KEEPIDLE, [30], 4) = 0
```
So in theory, it should detect outages when the probe fails. It doesn't set TCP_KEEPCNT, by default that is set to 9. So 4 minutes max outage before the keepalive probes should notice.

So I'm really baffled how it took an hour for one of my nodes to notice it lost the connect. Maybe it's a more fundamental race inside client-go??

**Comment by jsravn** [comment]
So to summarise my thoughts so far:
- Using an ELB is problematic because ELB doesn't handle the long lived kubelet/kube-proxy connections well, causing node NotReady outages if the ELB scales up and down. This can be mitigated by one of: 1) swapping to DNS 2) using more lenient grace periods on the node controller, to give kubelet a chance to reconnect (shouldn't take more than 4-5 minutes, I would guess) 3) some active monitoring of ELB IPs, and restart the daemons if they change. I still think client-go should actively heartbeat to more aggressively detect this connection loss - it seems to rely on detecting FIN to reconnect, which won't happen in the case of a VM being torn down completely (ELB scale up / scale down).
- HA etcd/apiserver seems to have some of its own issues that can cause similar looking problems of stale watches/updates. I'm not sure what causes it yet though. My best guess so far would be read quorum being disabled by default in the apiservers, so I'm going to try enabling it and see if it improves the situation.

**Comment by colemickens** [comment]
@jsravn that aligns pretty much exactly with what we were seeing here: https://github.com/kubernetes/kubernetes/issues/41185

**Comment by jeremyd** [comment]
Hi all!
The saga continues and I thought I'd ping back here with our current settings because they seem to be working pretty well.  So we had a few types of docker+kubelet flakes it seems.  

First issue would be our nodes would eventually be unable to start containers while remaining Ready and when logging into the worker docker ps would hang.  These nodes had to be killed off and we couldn't find concrete cause so we added a docker-healthcheck similar to what kops uses.  code here: https://gist.github.com/jeremyd/fc27e94d05d6e76cf7f26e41a5be4c31  The AMAZING part about this is that since we started running this our workers have stopped having the problem!  Our theory is that running docker ps periodically helps prevent docker from going out to lunch.

Second issue was that we had flurries of nodes going NotReady for no real reason that we could tell.  Since then we've increased NodeMonitorGracePeriod to 300s and NodeStatusUpdateFrequency to 20s.  The theory I have is that NodeStatusUpdate is very heavy on running it's readiness checks as it has to collect all stats involved with that node every period.  We also started running our cluster with *plenty of headroom on memory.  We do 60/40 deployments and keeping the cluster under 50% utilized.  With all these changes our cluster is running really smoothly now and the nodeNotReady flurries have subsided.

Kube 1.6.4, etcd3, flannel 0.7.1, docker 1.12.6, CoreOS Stable 1353.7.0

**Comment by jsravn** [comment]
Quick update: We haven't had any issues w/ this since swapping away from ELBs to DNS round robin, and enabling quorum reads. But still think client-go should handle ELBs better :).

**Comment by obeattie** [comment]
This can be fairly easily reproduced:

1. Start the `kubelet` and allow it to go `Ready`
2. Insert iptables rules to drop packets destined for the apiserver:
  ```bash
iptables -I OUTPUT 1 -m mark --mark 0x8 -j LOG --log-prefix "drop for apiserver: " --log-level 4
iptables -I OUTPUT 2 -m mark --mark 0x8 -j DROP
iptables -I OUTPUT 1 -d "$master_cidr" -p tcp --dport 443 -j MARK --set-mark 0x8
```
3. Observe the node transition to `NotReady` as reported by the apiserver
4. After ~20 minutes, the `kubelet` will log a lot of things like
  ```
E0629 10:37:24.595834    8173 kubelet_node_status.go:302] Error updating node status, will retry: error getting node "ip-10-101-36-7.eu-west-1.compute.internal": Get https://k8s-master.eu-west-1.i.s101.nonprod-ffs.io/api/v1/nodes?fieldSelector=metadata.name%3Dip-10-101-36-7.eu-west-1.compute.internal&resourceVersion=0: read tcp 10.101.36.7:55830->10.101.27.132:443: read: connection timed out
```

**Comment by obeattie** [comment]
Having investigated this a little further, I think I now understand this behaviour fully. 🔍 

Because the Kubelet does not perform application-layer heartbeating, the only way it knows the connection has been lost is when the kernel decides so. As mentioned above, Kubernetes does configure keepalives but these aren't at play here since they are only sent after 2 hours of inactivity by default ([`net.ipv4.tcp_keepalive_time`](https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt)).

Instead, TCP retransmissions account for the 15 minute delay. The default value for [`net.ipv4.tcp_retries2`](https://www.kernel.org/doc/Documentation/networking/ip-sysctl.txt) means that with exponential backoff, it takes ≈15m25s before the kernel kills the connection. Kubernetes places no timeout on its HTTP requests so the "update node status" request hangs for this time. When the 15 minutes expire, all the requests fail and Kubernetes reconnects. This situation is not entirely specific to the use of an ELB – if a directly-connected apiserver becomes partitioned or dies ungracefully, it will take 15 minutes before the Kubelet attempts to reconnect.

I can see several solutions here:

1. Instruct users to tune their sysctl's for the Kubelet's network namespace so it detects connection loss sooner. This doesn't feel like a practical solution to me but it's possible and does work in my tests.
2. Set options like `TCP_USER_TIMEOUT` on the TCP socket to more aggressively close the connection when packets sent to the apiserver remain unacknowledged. I've written some code to do this on a custom build of the Kubelet and it works well. However, this would introduce platform-dependent behaviour: `TCP_USER_TIMEOUT` only exists on Linux (there is `TCP_MAXRT` on Windows and probably equivalents on other platforms too, but it is unlikely to be supported on all systems).
3. Introduce application-level pings and terminate the connection if a number of consecutive pings fail. In my mind this is the "best" option but also the most complex – a ping endpoint would need to be added to the apiserver along with Kubelet code to call it and tear down the connection if it fails. Simply making a request to `/` would work, but all the 404's in the apiserver logs would not be intuitive.

I am happy to submit a patch for either option 2 or 3 but would appreciate some input on which would be preferred 🙂

**Comment by redbaron** [comment]
What about option 4. - introduce request timeout on the client side? If kubelet suffers from this defect, then all other users of client-go might be affected as well.

**Comment by obeattie** [comment]
A request-level timeout will not help here as the problem is at the connection level. I just added a request timeout and tested the behaviour. While individual requests to `/api/v1/nodes` fail after the timeout, the underlying connection (which is shared among all requests) remains active until the kernel declares it's dead. Subsequent requests get mapped to the same connection and continue to fail until the 15 minutes are up.

Without adding code to the `transport` package (which is currently the only package which has access to the connection itself) or configuring the kernel this can't be dealt with "higher up the call stack."

I personally would lean toward option 2, but the platform-dependent code is not ideal.

**Comment by obeattie** [comment]
A request-level timeout will not help here as the problem is at the connection level. I just added a request timeout and tested the behaviour. While individual requests to /api/v1/nodes fail after the timeout, the underlying connection (which is shared among all requests) remains active until the kernel declares it's dead. Subsequent retries get mapped to the same connection and continue to fail until the 15 minutes are up.

Without adding code to the transport package (which is currently the only package which has access to the connection itself) or configuring the kernel this can't be dealt with "higher up the call stack."

I personally would lean toward option 2, but the platform-dependent code is not ideal.

**Comment by redbaron** [comment]
I am not familiar with golang http client code, but are you saying that it is structured in a such way, that user of a it can't close connection in case of an error?

**Comment by obeattie** [comment]
No, I'm not saying that it is impossible to do with Go. I'm saying that the Go `net/http` code as well as the k8s `client-go` APIs are structured such that low-level connection details are abstracted from callers. Of course these APIs can be changed, but I think these connection-level concerns are best dealt with inside the transport package, where the necessary objects are already available.

Is your proposal that in the case any request times out, we close the connection?

**Comment by redbaron** [comment]
So there is some layer which abstracts connection, therefore it is responsible for establishing connection and maintaining it in a healthy state.  How to achieve that is an interesting question, given that you said that connection is shared among other higher level users and therefore closing it on timeout of one request might not be desirable, since timeout can be happening for reasons other than stalled TCP connection.

**Comment by Dao007forever** [comment]
@obeattie Would you mind posting your patch of TCP_USER_TIMEOUT? Did you change only `kubernetes/staging/src/k8s.io/client-go/rest/request.go`?

**Comment by derekwaynecarr** [comment]
@obeattie - thank you for your analysis here:
https://github.com/kubernetes/kubernetes/issues/41916#issuecomment-312428731

there is a similar issue I opened which is independent of any cloud provider:
https://github.com/kubernetes/kubernetes/issues/48638

@gmarek and I discussed having separate timeouts for watch vs non-watch operations.

i don't see a specific knob to set the tcp retry timeouts in go library, but i may have missed:
https://golang.org/pkg/net/http/#Transport

i like the idea of recommending to tweak the tcp retry values so its in the range between node status update monitoring interval (10s), and the node controller observing the node as lost (5m).  we set some sysctls in kubelet startup today already.

**Comment by derekwaynecarr** [comment]
this is not AWS specific issue, i am closing this issue and will track it in #48638
